src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2581,7 +2581,19 @@ internal DiagnosticBag AdditionalCodegenWarnings
             }
         }
 
+        /// <summary>
+        /// A bag in which circular struct diagnostics should be reported.
+        /// </summary>
+        internal BindingDiagnosticBag CircularStructDiagnostics
+        {
+            get
+            {
+                return _circularStructDiagnostics;
+            }
+        }
+
         private readonly DiagnosticBag _additionalCodegenWarnings = new DiagnosticBag();
+        private readonly BindingDiagnosticBag _circularStructDiagnostics = new BindingDiagnosticBag();
```

Comment:
>new BindingDiagnosticBag() [](http://example.com/codeflow?start=75&length=26)

Is this a "```BindingDiagnosticBag``` **with dependencies bag backed by a concurrent set**"?

Reply:
@AlekseyTs Thanks for the pointer. I assume it's okay to diverge a little bit from line 2746 as long as I'm using a `ConcurrentSet<AssemblySymbol>`? Specifically, Can I use `new DiagnosticBag()` instead of `DiagnosticBag.GetInstance()` to avoid knowing when to return it back to the pool?

If using the pool is the preferred approach, let me know if freeing after this line is correct:

https://github.com/dotnet/roslyn/blob/0b3dcee9d591e98374a8d341cba4a94c5f10dece/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs#L207

Reply:
> Can I use ```new DiagnosticBag()``` instead of ```DiagnosticBag.GetInstance()``` to avoid knowing when to return it back to the pool?

Yes, this is the way to go, we usually do not use pool for long-lived instances.